elastic/kibana
# [ResponseOps][actions] add config for allow-listing email address domains

## Summary

resolves https://github.com/elastic/kibana/issues/126944

docs preview:

- [connector settings](https://kibana_129001.docs-preview.app.elstc.co/guide/en/kibana/master/alert-action-settings-kb.html#action-settings)
- [email connector](https://kibana_129001.docs-preview.app.elstc.co/guide/en/kibana/master/email-action-type.html)

Adds a new configuration setting for the `actions` plugin, `xpack.actions.email.domain_allowlist`, which is an array of domain name strings which are allowed to be sent emails by the email connector.  If this setting isn't used, there is no change to the way the email connector works before this PR - email addresses are generally only validated when the connector is executed, by `node-mailer`. If this setting is used, for each email connector, each email address specified in the `from` property in config, and `to`, `cc`, and `bcc` properties in params, will be validated, and their domains checked against the allow list.  If any email is invalid (so we can't determine the domain), or it's domain is not in the allow list, that's considered an error condition, when the validation is run:  

- a connector create will fail if an invalid/not allowed email is used in the config `from` property
- a rule create/update will fail if an invalid/not allowed email is used in the email connector params `to`, `cc`, or `bcc` of any of it's actions
- and a connector run will fail if any of `from`, `to`, `cc`, or `bcc` are invalid/not allowed.

This adds a new dependency on an npm package [`email-addresses`](https://www.npmjs.com/package/email-addresses), which we use to extract the domain from an email address.  This also validates the email address, first.  This seems like a well-maintained package and has no dependencies.

We expose the config setting down into the UX via `exposeToBrowser`.  A chat with Kibana Security indicated they didn't feel like this was a security concern.  Basically, if we **WANT** to be able to validate email addresses in the UX, we will have to expose **SOME** way of validating email addresses via HTTP, so even the most obscure way of doing this (for example, exposing a new server endpoint to validate email addresses) will end up "leaking" which domains are actually allowed.  If this is still felt to be a security concern, we can rip out everything on the UX side, and let the validation only happen on the server.  The same error conditions will work, but the UX is not quite as nice - "Error saving connector: blah blah" instead of nice error indicators in appropriate fields in the UX.

This required adding a new `public` plugin for `actions` - it didn't have one before.  The `common` folder for actions is also where the new validation logic is implemented, where it calls into the `email-addresses` package, so it's shared between client and server.  

Creating this new plugin caused some other issues, with some ServiceNow public code that was referencing the server plugin types, but using a `// eslint-disable-next-line @kbn/eslint/no-restricted-paths` comment to disable the compile time warning.  With the new public plugin, this changed to a runtime warning message.  So, the relevant types were extracted into common, and then the references to them changed.  Seems like a good cleanup, and eliminated some `//eslint-disable` usages.

For function tests, `security_and_spaces` was left alone, so is the tester for **NOT** using the new config.  `spaces_only` now uses the new setting, and has added tests to ensure the code works as expected.

### To Verify

Add the new config, for example:

    xpack.actions.email.domain_allowlist: ['elastic.co']

Then try creating a connector with a `from` with a different email address, and with an `elastic.co` address.  Should only work for `elastic.co`.  After creating a connector, try using the test page and using different values for `to`, `cc`, and `bcc`.  Should only work with `elastic.co` addresses.  Then add the email connector to an existing rule, and try setting the `to`, `cc`, and `bcc` to different values to see they work or generate errors.  

Note, the email doesn't have to be a "live" one, you can use Gmail but give it bogus userid / password.  In which case if you see a failure from Gmail, then the email _would_ have validated and sent :-).

You can also test using previously created email connectors without the setting, then turn the setting on, and edit the connector to see the "new" errors.

### Checklist

Delete any items that are not applicable to this PR.


## Release Note
Adds a new configuration setting for the `actions` plugin, `xpack.actions.email.domain_allowlist`, which is an array of domain name strings which are allowed to be sent emails by the email connector.
